lib/view_component/base.rb

```diff
@@ -284,20 +305,36 @@ def template_errors
             unless invalid_variants.empty?
               errors << "More than one template found for #{'variant'.pluralize(invalid_variants.count)} #{invalid_variants.map { |v| "'#{v}'" }.to_sentence} in #{self}. There can only be one template file per variant."
             end
+
+            if templates.find { |template| template[:variant].nil? } && inline_calls_defined_on_self.include?(:call)
+              errors << "Template file and inline render method found for #{self}. There can only be a template file or inline render method per component."
+            end
+
+            duplicate_template_file_and_inline_variant_calls =
+              templates.pluck(:variant) & variants_from_inline_calls(inline_calls_defined_on_self)
+
+            unless duplicate_template_file_and_inline_variant_calls.empty?
+              count = duplicate_template_file_and_inline_variant_calls.count
+
+              errors << "Template #{'file'.pluralize(count)} and inline render #{'method'.pluralize(count)} found for #{'variant'.pluralize(count)} #{duplicate_template_file_and_inline_variant_calls.map { |v| "'#{v}'" }.to_sentence} in #{self}. There can only be a template file or inline render method per variant."
+            end
+
             errors
           end
       end
 
-      def compiled_template(file_path)
-        handler = ActionView::Template.handler_for_extension(File.extname(file_path).gsub(".", ""))
-        template = File.read(file_path)
-
-        if handler.method(:call).parameters.length > 1
-          handler.call(self, template)
-        else
-          handler.call(OpenStruct.new(source: template, identifier: identifier, type: type))
+      def variants_from_inline_calls(calls)
+        calls.reject { |call| call == :call }.map do |variant_call|
+          variant_call.to_s.sub("call_", "").to_sym
         end
       end
+
+      def view_component_ancestors
```

Comment:
Is there a reason for this to be its own method? Could we inline it above instead? (keeping the nice comment you've written here, of course ❤️ )